### PR TITLE
feat: Obtain repository files from tarball

### PR DESCRIPTION
**Pull request by zharinov**
## Changes

Proof of concept using `.tar.gz` on-the-fly decompression

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/platform/github/scm.ts [hunk 1]
```diff
@@ -1,7 +1,18 @@
+import gunzip from 'gunzip-maybe';
+import tar from 'tar-stream';
```
Comment (viceice) [inline lib/modules/platform/github/scm.ts:2]: I'm using node-tar on containerbase

https://www.npmjs.com/package/tar

File: lib/modules/platform/github/scm.ts [hunk 2]
```diff
@@ -11,4 +22,34 @@ export class GithubScm extends DefaultGitScm {
...
+    const extractTar = tar.extract({
+      filenameEncoding: 'utf-8',
+      allowUnknownFormat: true,
+    });
+    const stream = githubApi
+      .stream(tarballUrl)
+      .pipe(extractGzip)
+      .pipe(extractTar);
+
+    const files: string[] = [];
```
Comment (viceice) [inline lib/modules/platform/github/scm.ts:41]: currently unused.

**Review by viceice — COMMENTED** [review]
what will be the benefit? 🤔

**Comment by zharinov** [comment]
> what will be the benefit? 🤔

If done right, it's one more step to complete repo processing without the cloning

**Comment by rarkins** [comment]
But it downloads the full files of the default branch?

**Comment by zharinov** [comment]
But instead of just listing files, it should filter it on-the-fly and select the content of those which are needed. Yes, for just file listings it's not that useful.

**Comment by zharinov** [comment]
> But it downloads the full files of the default branch?

It doesn't *download* full arhive to the memory or disk, it processes it chunk by chunk while network request goes. Something like file hash calculation for some of our datasources.

**Comment by rarkins** [comment]
What scenarios is this more efficient? Because if any commit is made then a clone is still needed right?

**Comment by zharinov** [comment]
BTW, it's not only for default branches, you may point to any other base branch

> What scenarios is this more efficient? Because if any commit is made then a clone is still needed right?

1. New commits to the repo are made, Renovate extracts relevant files, checks dependencies, finishes if there is no updates (no git involved)
2. (probably, in the future) If updates are available with no artifacts involved, perform API-based commits

**Comment by viceice** [comment]
but is it really more performant than cloning with blob=none filter? 🤔

**Pull request closed without merging**